File: src/SIL.Machine/Translation/WordEditDistance.cs
using System.Collections.Generic;

namespace SIL.Machine.Translation
{
	internal class WordEditDistance : EditDistance<string, char>
	{
		public double HitCost { get; set; }
		public double InsertionCost { get; set; }
		public double DeletionCost { get; set; }
		public double SubstitutionCost { get; set; }

		public double Compute(string x, string y, out IEnumerable<EditOperation> ops)
		{
			double[,] distMatrix;
			double dist = ComputeDistMatrix(x, y, true, false, out distMatrix);
			ops = GetOperations(x, y, distMatrix, true, false, GetCount(x), GetCount(y));
			return dist;
		}

		public double ComputePrefix(string x, string y, bool isLastItemComplete, bool usePrefixDelOp,
			out IEnumerable<EditOperation> ops)
		{
			double[,] distMatrix;
			double dist = ComputeDistMatrix(x, y, isLastItemComplete, usePrefixDelOp, out distMatrix);
			ops = GetOperations(x, y, distMatrix, isLastItemComplete, usePrefixDelOp, GetCount(x), GetCount(y));
			return dist;
		}

		protected override int GetCount(string item)
		{
			return item.Length;
		}

		protected override char GetItem(string seq, int index)
		{
			return seq[index];
		}

		protected override double GetHitCost(char x, char y, bool isComplete)
		{
			return HitCost;
		}

		protected override double GetSubstitutionCost(char x, char y, bool isComplete)
		{
			return SubstitutionCost;
		}

		protected override double GetDeletionCost(char x)
		{
			return DeletionCost;
		}

		protected override double GetInsertionCost(char y)
		{
			return InsertionCost;
		}

		protected override bool IsHit(char x, char y, bool isComplete)
		{
			return x == y;
		}
	}
}

File: src/SIL.Machine.TS/src/translation/word-edit-distance.ts
import { EditDistance } from './edit-distance';
import { EditOperation } from './edit-operation';

export interface WordEditDistanceResult {
  cost: number;
  ops: EditOperation[];
}

export class WordEditDistance extends EditDistance<string, string> {
  hitCost: number = 0;
  insertionCost: number = 0;
  deletionCost: number = 0;
  substitutionCost: number = 0;

  compute(x: string, y: string): WordEditDistanceResult {
    const matrixResult = this.computeDistMatrix(x, y, true, false);
    const ops = this.getOperations(x, y, matrixResult.distMatrix, true, false, this.getCount(x), this.getCount(y));
    return { cost: matrixResult.cost, ops };
  }

  computePrefix(x: string, y: string, isLastItemComplete: boolean, usePrefixDelOp: boolean): WordEditDistanceResult {
    const matrixResult = this.computeDistMatrix(x, y, isLastItemComplete, usePrefixDelOp);
    const ops = this.getOperations(
      x,
      y,
      matrixResult.distMatrix,
      isLastItemComplete,
      usePrefixDelOp,
      this.getCount(x),
      this.getCount(y)
    );
    return { cost: matrixResult.cost, ops };
  }

  protected getCount(seq: string): number {
    return seq.length;
  }

  protected getItem(seq: string, index: number): string {
    return seq[index];
  }

  protected getHitCost(x: string, y: string, isComplete: boolean): number {
    return this.hitCost;
  }

  protected getSubstitutionCost(x: string, y: string, isComplete: boolean): number {
    return this.substitutionCost;
  }

  protected getDeletionCost(x: string): number {
    return this.deletionCost;
  }

  protected getInsertionCost(y: string): number {
    return this.insertionCost;
  }

  protected isHit(x: string, y: string, isComplete: boolean): boolean {
    return x === y;
  }
}
